File: backend/routers/notifications.py
"""Notifications router — read/mark-read notification records."""

from datetime import datetime, timezone
from typing import Annotated
from fastapi import APIRouter, Depends, Query
from backend.deps import require_auth
from backend.core.db import fetch_one, fetch_all, execute as db_execute

router = APIRouter(tags=["notifications"])
Usr = Annotated[dict, Depends(require_auth)]


@router.get("")
def get_notifications(user: Usr, unread_only: bool = Query(False), limit: int = Query(50)):
    where = "nr.user_id = ?"
    params: list = [user["id"]]
    if unread_only:
        where += " AND nr.read_at IS NULL AND nr.dismissed_at IS NULL"
    rows = fetch_all(
        f"""SELECT n.*, nr.read_at, nr.dismissed_at
            FROM notifications n
            JOIN notification_recipients nr ON nr.notification_id = n.id
            WHERE {where}
            ORDER BY n.created_at DESC
            LIMIT ?""",
        params + [limit],
    )
    return rows


@router.get("/unread-count")
def unread_count(user: Usr):
    row = fetch_one(
        """SELECT COUNT(*) AS n FROM notification_recipients
           WHERE user_id=? AND read_at IS NULL AND dismissed_at IS NULL""",
        (user["id"],),
    )
    return {"count": row["n"] if row else 0}


@router.post("/{notif_id}/read")
def mark_read(notif_id: int, user: Usr):
    db_execute(
        "UPDATE notification_recipients SET read_at=? WHERE notification_id=? AND user_id=? AND read_at IS NULL",
        (datetime.now(timezone.utc).isoformat(), notif_id, user["id"]),
    )
    return {"ok": True}


@router.post("/mark-all-read")
def mark_all_read(user: Usr):
    db_execute(
        "UPDATE notification_recipients SET read_at=? WHERE user_id=? AND read_at IS NULL",
        (datetime.now(timezone.utc).isoformat(), user["id"]),
    )
    return {"ok": True}

File: school-mis-app/src/api/notifications.ts
import api from "@/api/client";

export interface Notification {
  id:           number;
  type:         string;
  title:        string;
  body:         string;
  data_json:    string | null;
  created_at:   string;
  read_at:      string | null;
  dismissed_at: string | null;
}

export async function getNotifications(unreadOnly = false, limit = 50): Promise<Notification[]> {
  const r = await api.get("/notifications", { params: { unread_only: unreadOnly, limit } });
  return r.data;
}

export async function getUnreadCount(): Promise<number> {
  const r = await api.get("/notifications/unread-count");
  return r.data.count;
}

export async function markRead(id: number): Promise<void> {
  await api.post(`/notifications/${id}/read`);
}

export async function markAllRead(): Promise<void> {
  await api.post("/notifications/mark-all-read");
}

export async function getAuditLog(params?: {
  table_name?: string;
  actor_id?:   number;
  action?:     string;
  limit?:      number;
}): Promise<any[]> {
  const r = await api.get("/audit", { params });
  return r.data;
}
